### ISSUE TITLE
Add an easy way to get the instantiated layout for a @Route (Feature Request)

### ISSUE THREAD

**Issue by petrixh** [vaadin/flow#5443]
Could we have a way to get the instantiated RouterLayouts accessible somehow through a convenient API. For instance if you make a RouterLayout for the "Menu and Current user details" for an application, there are times when you'll need to access the "LayoutWithMenu" instance that is used for the routes. 

Currently the "best way" I could figure out is to override onAttach() on the components that has @Route(value="theView", layout=LayoutWithMenu.class) and use getParent() to store the parent instance in the route instance for later access. This is also a bit error prone as it'll not be populated before attach, which would guide developers to "do work" in the attach method which in Vaadin 8 at least was discouraged. 

Haven't thought the API completely through, but i was expecting to find something like: 
RouterLayout layout = ...getRouter().getCurrentRouterLayout();

or in plural as the layout-parameter in @Route can take multiple routes (maybe both as the single layout case is probably the 90% use-case and the singular version could always return the "first")

Vaadin Platform 10.0.4

**Comment by Legioth** [comment]
>  which would guide developers to "do work" in the attach method which in Vaadin 8 at least was discouraged

The problem with `attach()` in Vaadin 8 and older is that you must remember to call `super.attach()` and that weird things will happen with the component hierarchy if that methods throws an exception. 

These problems should be fixed starting from Vaadin 10 where `onAttach(AttachEvent)` is only a notification about what has happened, but there isn't any super implementation that would e.g. update internal bookkeeping.

**Comment by petrixh** [comment]
Good point Leif, that was part of the problem. Another was that as components are reused they are attached and detached several times which, if the initialisation logic is in the attach, causes the view to be initialised several times. Of course after realising this, one add a if(initialised)-flag but this is already starting to get messy...

**Comment by pleku** [comment]
@petrixh the `BeforeEnterEvent` and `BeforeLeaveEvent` have an API for `List<Class<? extends RouterLayout>> getLayouts()` which could be used to discover the layout.

I'm trying to figure out where and how one would like to obtain the layout, I personally don't like `RouterLayout layout = ...getRouter().getCurrentRouterLayout();`, how about `getRouter().getParentLayout(this);` where the `this` would be the route you have ?

**Comment by petrixh** [comment]
@pleku `getRouter().getParentLayout(this)` would be fine, would be really nice if you can still call it before the actual enter/afterNavigation events as sometimes it makes sense to init their view components and state early and simply populate/refresh some parts of them in the enter/after navigation methods. I understand that this might be hard, but at some point the FW will have to figure out the hierarchy anyways and (I'm probably grossly oversimplifying it) this could be done when someone (developer or FW) needs that info for the first time the classical: 

      public XYZ getParentLayouts(Component component){

          if(parentLayouts == null){
              parentLayouts = loadParents(component); 
          }

          return parentLayouts; 
      } 

The first UC that comes to mind is for a **non-spring** application developer that uses the parent layouts to construct the navigation structure/header of the application. Say that a view has a function that needs to temporarily hide the navigation bar, or based on the users selection in the view hides/shows some additional navigation entries/something in the header (crud buttons?). Now, if this happens "per user interaction" (i.e. user clicked on something) then you could have a stored instance of the parent layout (if you knew about the enter event thing). However, if the users access rights/application state or the URL params dictate this behaviour it becomes harder. In it's current form, you'd have to store the information on what to do and schedule it for a later time (like attach or beforeEnter), and most likely end up with copied code that 99% does the same thing, one just does it later... 

Another example, which is relevant in a current project, is that based on the loaded view, we want to show information related to the "current thing" in the side-bar navigation. Now, we can (and do) have the sidebar pull the info from the URL, but this means that now we have to "load the thing" twice from the db, and the second time it's just to pull the name from it... And we of course have to duplicate all the exception handling in both the view and the nav-bar (in case the user tampers with the URL, entity was deleted etc...) Would be nicer to be able to have the view load the entity, fetch the "parent navigation layout" and then "give" the loaded entity to it. We could store the value in the session, yes, and/or even use Spring to inject the thing/parent layout... but that was not the point of the original ticket :) 

The original ticket was for a KISS approach where you explicitly do **not** use Spring/CDI. Since this ticket was created I've found several "workarounds" for this, but these are the "symptoms", the "cause" should still be addressed so that no "symptomatic workarounds" are needed :)

**Comment by petrixh** [comment]
And one other thing to mention is that in bigger applications this would probably be done in an abstract superclass, so even if you had spring, you might not know the ”type” of the parent layouts, just a common interface as one implementation of a view might use one parent layout while another might use a different one...

For instance you have ”UserViewX” that uses ”NavRoot”, and ”SuperUserView” that uses ”SuperNavRoot” which extends from ”NavRoot”... now in both cases the actual view would probably extend from a common ”AbstractView” where some rights based things are done. So ”AbstractView” would need the current instance of ”NavRoot”, regardless if its the ”normal” or ”super user” variant...

**Comment by rolfsmeds** [comment]
Just to give a practical example of a case where this might come in handy:
Let's say you have a UI with a parent layout with navigation etc, but also e.g. a toolbar layout for things like _"Add x"_ or _"Edit"_ buttons (for adding a new item to the current view, or switching the current view to edit mode, respsectively). The design may require them to be in an area of the UI that, technically, is part of the parentlayout. So you want the current view to be able to easily utilize that area for adding those components.
